pkg/storage/db.go

```diff
@@ -30,6 +30,22 @@ const commonQuery = "SELECT " +
 	"	ON aws_resources_id = aws_resources.id " +
 	"WHERE "
 
+// can't use Sprintf in a const, so...
+// %s throughout should be `aws_hostnames_hostname` or `aws_ips_ip`
+const latestStatusQuery = "SELECT aws_events_ips_hostnames.aws_resources_id, aws_events_ips_hostnames.aws_ips_ip, aws_events_ips_hostnames.aws_hostnames_hostname, is_public, is_join, aws_events_ips_hostnames.ts, aws_resources.account_id, aws_resources.region, aws_resources.type, aws_resources.meta " +
```

Comment:
I'd leave it to the optimizer/scheduler to determine the truth of the matter, but I believe this query will result in several scans/nested loops because of the organization. I believe we can avoid them by restructuring the query with common table expressions and window functions. For example:

```sql
// Assuming:
// $1 = hostname
// $2 = timestamp requested
//
// Compute a filtered set of results that each has an attached "most recent ts"
// that is up to but not exceeding the given timestamp.
WITH latest_candidates AS (
    SELECT
        *,
        // the max is applied after the WHERE so the dataset to aggregate is much smaller.
        MAX(ts) OVER (PARTITION BY aws_events_ips_hostnames.aws_resources_id) as max_ts
    FROM aws_events_ips_hostnames
    WHERE
        aws_events_ips_hostnames.hostname = $1 AND
        aws_events_ips_hostnames.ts <= $2
),
// Drop all but only the most recent matching records for each resource
// that would have been the truth at the requested time.
latest AS (
    SELECT *
    FROM latest_candidates
    WHERE latest_candidates.ts = latest_candidates.max_ts
)
SELECT
    latest.*,
    aws_resources.account_id,
    aws_resources.region,
    aws_resources.type,
    aws_resources.meta
FROM latest
LEFT OUTER JOIN
    aws_resources ON 
        latest.aws_resources_id = aws_resources.id
```

I'm pulling from a fairly cold part of my SQL knowledge cache so the syntax may not be 100% correct. However, this is a general strategy I've seen work well for removing nested sub-queries and, often as a result, full table scans and nested loops from the optimizer plans. It also removes the need for a second timestamp by removing the need to do any kind of range query.

One thing that I'm definitely not sure about is how partitioning affects all of this.

Reply:
Great!  thanks!  I'll work through these improvements.

Reply:
Incorporated your suggested changes.  Thanks again!